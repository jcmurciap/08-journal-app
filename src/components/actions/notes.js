// libraries
import Swal from 'sweetalert2';
import { collection, addDoc, doc, updateDoc, deleteDoc} from 'firebase/firestore';

// components
import { types } from '../types/types';
import { loadNotes } from "../../helpers/loadNotes";
import { fileUpload } from '../../helpers/fileUpload';
import { db } from "../../firebase/firebase-config";

export const startNewNote = () => {

    return async (dispatch , getState) => {
        
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await addDoc( collection( db, `${ uid }/journal/notes` ), newNote );
        
        dispatch( activeNote( doc.id, newNote ) );
        dispatch( addNewNote( doc.id, newNote ) );
    }
}

export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note,
    }
});

export const addNewNote = ( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note,
    }
});

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    }
};

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes,
});

export const startSaveNote = ( note ) => {
    
    return async( dispatch, getState ) => {
        
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        };

        const noteToFirestore = {...note}; // id,body,date,title,url
        delete noteToFirestore.id; // body,date,title,url

        // guarda en la base de datos
        await updateDoc(doc(db, `${uid}/journal/notes/${note.id}`), noteToFirestore);
    
        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success');
    }; 
};


// actualiza del store el titulo y el body de la tarjeta que cambia
export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note,
        }
    }
});  

export const startUploading = ( file ) => {
    
    return async( dispatch, getState ) => {
        
        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            willOpen: () => {
                Swal.showLoading();
            }
        });
        
        const fileUrl = await fileUpload( file );
        
        activeNote.url = fileUrl;
        dispatch( startSaveNote( activeNote ) );
        
        Swal.close();
    };
};

export const startDeleting = ( id ) => {
    
    return async( dispatch, getState ) => {
        
        const uid = getState().auth.uid;
        await deleteDoc( doc(db, `${ uid }/journal/notes/${ id }`));
        dispatch( deleteNote( id ) );
    };
};

export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id,
});

export const noteLogout = ( notes ) => ({
    
    type: types.notesLogoutCleaning,
})
