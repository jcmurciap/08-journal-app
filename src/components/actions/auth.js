// libraries
import Swal from 'sweetalert2';
import 'firebase/firestore';
import { getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { googleAuthProvider } from '../../firebase/firebase-config';

// components
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';
import { types } from '../types/types';

export const startLoginEmailPassword = ( email, password ) => {

    
    return ( dispatch ) => {
        // gateway to the Firebase authentication API
        const auth = getAuth();
        
        dispatch( startLoading() );
        
        return signInWithEmailAndPassword( auth, email, password )
            .then(({ user }) => {
                dispatch( login( user.uid, user.displayName ) );
                dispatch( finishLoading() );
            })   
            .catch( e => {
                dispatch( finishLoading() );
                Swal.fire( 'Error', e.message, 'error' );
            });
    }; 
}

// save data in firebase
export const startRegisterWithEmailPasswordName = ( name, email, password ) => {
    
    return ( dispatch ) => {
        const auth = getAuth();

        // autenticacion
        createUserWithEmailAndPassword( auth, email, password )
            .then( async({ user }) => {
                await updateProfile( user, { displayName: name } );
                dispatch( 
                    login( user.uid, user.displayName ) 
                );
            })
            .catch( e  => {
                Swal.fire( 'Error', e.message, 'error' );
            })
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        const auth = getAuth();
        signInWithPopup( auth, googleAuthProvider )
            .then( ({ user }) => {
                dispatch( login( user.uid, user.displayName ) );
            });
    };
}

export const login = ( uid, displayName ) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName 
        }
    };
}

export const startLogout = () => {    
    
    return async( dispatch ) => {
        
        const auth = getAuth();
        await signOut( auth );
        dispatch( logout() );
        dispatch( noteLogout() );  
    };
}

export const logout = () => ({
    type: types.logout
})
