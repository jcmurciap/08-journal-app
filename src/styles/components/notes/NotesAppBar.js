import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../../components/actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector( state => state.notes );
    
    const handleSave = () => {
        // active === note
        dispatch(startSaveNote( active ))
    }
    
    const handlePictureUpload = () => {
        document.querySelector('#fileSelector').click();
    
    }
    
    const handleFileChange = ( e ) => {
        const file = e.target.files[0];
        
        if ( file ){
            dispatch(startUploading(file))
        }  
    
    }
    
    
    return (

        <div className="notes__appbar">
            <span>21 septiembre 2021</span>

            <input
                id="fileSelector"
                type="file"
                name="file"
                style={{ display: 'none' }}
                onChange={ handleFileChange }
            />
            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureUpload }
                >
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={ handleSave }
                >
                    Save
                </button>
            </div>
        </div>
    )
}
