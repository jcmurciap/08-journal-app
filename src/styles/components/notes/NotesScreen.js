import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../../components/actions/notes';
import { useForm } from '../../../components/hooks/useForms';
import { NotesAppBar } from './NotesAppBar'


export const NotesScreen = () => {
    
    const dispatch = useDispatch();
    
    // referencia a la nota activa
    const { active: note } = useSelector(state => state.notes);
    const [ formValues, handleInputChange, reset ] = useForm( note );
    const { title, body } = formValues;
    
    const activeId = useRef( note.id );
    
    // el efecto se dispara si y solo si el 'id' cambia
    useEffect( () => {
        // evita loop infinito entre useSelector y useForm
        if ( note.id  !== activeId.current ){
            reset( note );
            // nuevo valor de 'activeId'
            activeId.current = note.id
        }
    }, [ note, reset ] );

    
    useEffect(() => {
        
        dispatch(activeNote(formValues.id, { ...formValues })); // ...formValues se desestructura en 'notes'       
    }, [ formValues, dispatch ])


    const handleDelete = () => {
        dispatch(startDeleting( note.id ));    
    }
    
    
    
    return (
        
        <div className="notes__main-content">
            <NotesAppBar />
            
            {/* Agrupador del formularo */}
            <div className="notes__content">
                <input 
                    className="notes__title-input"
                    placeholder="Some awesome title"
                    type="text"
                    name="title"
                    value={ title }
                    onChange={ handleInputChange }
                />
                
                <textarea
                    className="notes__textarea"
                    placeholder="What hapenned today"
                    name="body"
                    value={ body }
                    onChange={ handleInputChange }
                >
                </textarea>

                {
                    ( note.url ) &&
                    (
                        <div className="notes__image">
                            <img
                                src={note.url}
                                alt="code"
                            />
                        </div>
                    )
                }

            </div>
            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>    
        
        
        </div>
    )
}

