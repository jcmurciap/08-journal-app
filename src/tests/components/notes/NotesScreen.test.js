import React from 'react';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import { NotesScreen } from "../../../styles/components/notes/NotesScreen";
import configureStore from 'redux-mock-store';
import { activeNote } from '../../../components/actions/notes';
import { Provider } from 'react-redux';

jest.mock('../../../components/actions/notes', () => ({
    activeNote: jest.fn()
}));

const middlewares = [thunk];

// create a store
const mockStore = configureStore(middlewares);

// store current status
const initState = {
    auth: {
        uid: '1',
        name: 'camilo'
    },
    ui: {
        loading: true,
        msgError: null,
    },
    notes: {
        active: {
            id: 1234,
            title: 'hello',
            body: 'world',
            date: 0,
        },
        notes: [],
    }
};

let store = mockStore(initState);

// mock the store
store.dispatch = jest.fn(); 



describe( '<NotesScreen /> tests', () => {

    
    const wrapper = mount(
        <Provider store={store}>
            <NotesScreen/>
        </Provider>
    )
    
    test('show skeleton correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Should to dispatch activeNote action component', () => {

        // modifica la caja de texto
        wrapper.find('input[name="title"]').simulate('change', {
            // evento
            target: {
                name: 'title',
                value: 'hello again',
            } 
        });
        
        // activeNote fue llamado 2 veces,por el useEffect(al inicio y al actualizar)
        expect( activeNote ).toHaveBeenLastCalledWith(
            1234,
            {
                body: 'world',
                title: 'hello again',
                id:1234,
                date: 0
            }
        )
    });
});