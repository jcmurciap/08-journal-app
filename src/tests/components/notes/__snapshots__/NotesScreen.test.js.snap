// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotesScreen /> tests show skeleton correctly 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      21 septiembre 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hello"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What hapenned today"
      value="world"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
